File: 2201-count-artifacts-that-can-be-extracted/2201-count-artifacts-that-can-be-extracted.py
class Solution:
    def digArtifacts(self, n: int, artifacts: List[List[int]], dig: List[List[int]]) -> int:
        digSet = set()
        
        for each in dig:
            digSet.add(tuple(each))
            
        ans = 0
        
        for point in artifacts:
            point1 = point[ : 2]
            point2 = point[2 : ]
            
            completeVisit = True
            
            for x in range(point1[0], point2[0] + 1):
                for y in range(point1[1], point2[1] + 1):
                    if (x, y) not in digSet:
                        completeVisit = False
            if completeVisit:
                ans += 1
        
        return ans

File: 2201-count-artifacts-that-can-be-extracted/2201-count-artifacts-that-can-be-extracted.cpp
class Solution {
public:
    int digArtifacts(int n, vector<vector<int>>& artifacts, vector<vector<int>>& dig) {
        return solve(n, artifacts, dig);
    }
private:
    // Aakash sir
    int solve(int n, vector<vector<int>>& artifacts, vector<vector<int>>& dig) {
        set<pair<int, int>> digSet;
        
        // Make set for given pairs to have a constant lookup for any pair
        for(auto each: dig) {
            digSet.insert({each[0], each[1]});
        }
        
        int ans = 0;
        
        // iterate through given artifacts
        for(auto points: artifacts) {
            
            // assume that complete cell can be visited
            bool completeVisit = true;
            
            // iterate through every rectangle for each artifact
            for(int x = points[0]; x <= points[2]; ++x) {
                for(int y = points[1]; y <= points[3]; ++y) {
                    
                    // if any cell in the rectangle cannot be visited
                    // then break immediately
                    if(digSet.find({x, y}) == digSet.end()) {
                        completeVisit = false;
                        break;
                    }
                }
            }
            // if the artifact rectangle can be completely visited
            if(completeVisit)
                ++ans;
        }
        return ans;
    }
};